spec/java/util/stream/Stream.main.lsl

```diff
@@ -0,0 +1,1049 @@
+    }
+
+
+    // maxSize: long - what to do with this ? Array
+    fun *.limit (@target self: Stream, maxSize: long): Stream
+    {
+        if (maxSize < 0)
+            action THROW_NEW("java.lang.IllegalArgumentException", []);
+
+        if (maxSize == 0)
+        {
+            result = new StreamAutomaton(state = Initialized,
+                storage = this.storage,
+                length = 0,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+        // Maybe only change length field ? And don't change storage ?
+        else
+        {
+            // what will be if will be overflow ?
+            val maxSizeInt: int = maxSize as int;
+            val limitStorage: array<Object> = action ARRAY_NEW("java.lang.Object", maxSizeInt);
+
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 0, maxSizeInt, +1,
+                _limit_loop(i, limitStorage)
+            );
+
+            result = new StreamAutomaton(state = Initialized,
+                storage = limitStorage,
+                length = maxSizeInt,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+    }
+
+
+    @Phantom proc _limit_loop (i: int, limitStorage: array<Object>): void
+    {
+        limitStorage[i] = this.storage[i];
+    }
+
+
+    fun *.skip (@target self: Stream, n: long): Stream
+    {
+        if (n < 0)
+            action THROW_NEW("java.lang.IllegalArgumentException", []);
+
+        if (n == 0)
+        {
+            result = new StreamAutomaton(state = Initialized,
+                storage = this.storage,
+                length = this.length,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+        else
+        {
+            // what will be if will be overflow ?
+            val offset: int = n as int;
+            val newLength: int = this.length - offset;
+            val skipStorage: array<Object> = action ARRAY_NEW("java.lang.Object", newLength);
+
+            var i: int = 0;
+            var skipIndex: int = 0;
+            action LOOP_FOR(
+                i, offset, this.length, +1,
+                _skip_loop(i, skipIndex, skipStorage)
+            );
+
+            result = new StreamAutomaton(state = Initialized,
+                storage = skipStorage,
+                length = newLength,
+                closeHandlers = this.closeHandlers,
+            );
+        }
+    }
+
+
+    @Phantom proc _skip_loop (i: int, skipIndex: int, skipStorage: array<Object>): void
+    {
+        skipStorage[skipIndex] = this.storage[i];
+        skipIndex += 1;
+    }
+
+
+    fun *.forEach (@target self: Stream, _action: Consumer): void
+    {
+        _actionApply(_action);
+    }
+
+
+    fun *.forEachOrdered (@target self: Stream, _action: Consumer): void
+    {
+        _actionApply(_action);
+    }
+
+
+    fun *.toArray (@target self: Stream): array<Object>
+    {
+        result = this.storage;
```

Comment:
Streams are generally a one-time-use objects. How it would be modeled here?
Consider the following use case:
```java
var s = new Random(123).ints(5).limit(20);
var a = s.toArray();
var b = s.toArray();  // !!!
```
The original `Stream` would throw an exception: `IllegalStateException: stream has already been operated upon or closed`.

Reply:
@dog-m Maybe add flag "arrayWasCreated" (for example) ? If it was true - throw IllegalStateException ? 
What do you think ?

Reply:
I suggest adding a `var closed: boolean = false;` internal automaton variable and **updating and checking** it in every(?) *chaining* (that what allows chaining methods together) and *terminal* (that what returns a final processed value: an array, a value, etc.) method.